Improve metric and generate plot for in-flight batch size benchmark (#422)
triton-inference-server/client: src/c++/perf_analyzer/docs/llm.md

tokens in order to increase the in-flight batch size over time.

#### Example

In this benchmark, we will run Perf Analyzer in [periodic concurrency mode](inference_load_modes.md#periodic-concurrency-mode)
that periodically launches a new concurrent request to the model using `--periodic-concurrency-range START END STEP` option.
In this example, Perf Analyzer starts with a single request and launches the new ones until the total number reaches 100.
You can also specify the timing of the new requests: Setting the `--request-period` to 32 (as shown below) will make
Perf Analyzer to wait for all the requests to receive 32 responses before it launches the new requests.
Run the following command inside the client container.

```bash
# Install matplotlib to generate the benchmark plot
pip install matplotlib

# Run Perf Analyzer
python profile.py -m vllm --periodic-concurrency-range 1 100 1 --request-period 32 --max-tokens 1024 --ignore-eos

# Sample output
# Saved benchmark result @ 'inflight_batching_benchmark.png'.
```

The resulting plot will look like

<img src="examples/inflight_batching_benchmark.png" width="600">

> The plot above shows the change in average token-to-token latency at each request period sized segments across the benchmark.

Reviewer: Nit: add clarity. "each request period sized segments" is a bit confusing.

What this means is that we split the entire benchmark timeline into segments, each with equal size of `--request-period`.

> In this example, the first segment will contain 32 responses, the second segment will contain 2*32 responses, and so on:

Reviewer: ```suggestion
In this example, the first segment will contain 32 responses,
the second segment will contain 2*32 responses, and so on:
```

```
              32 responses (=request period)
                ┌────┐
      request 1 ──────┊──────┊──────┊──────┊──────┊
      request 2       ┊──────┊──────┊──────┊──────┊──────┊
      request 3       ┊      ┊──────┊──────┊──────┊──────┊──────┊

>       request 4       ┊      ┊      ┊──────┊──────┊──────┊──────┊──────

Reviewer 2: Amazing visual!!!

                      ┊      ┊      ┊      ┊      ┊      ┊      ┊
         ...

       segment    1      2      3       4      5      6      7      8      ...
(i-th request period)
```
For each segment (or i-th request period), we compute average token-to-token latency and plot the results.

> Since we send total 100 requests, expect 1024 responses back from server, and set request period to 32,

Reviewer: ```suggestion
We sent 100 requests and expect 1024 responses back from server.
We set the request period to 32, so there are 131 segments.
See [profile.py](examples/profile.py) for more details.
```

> **Note**
>
> This benchmark relies on the feature that will be available from `23.10`
> release which is on its way soon. You can either wait until the `23.10`
> container is ready or build Perf Analyzer from the latest `main` branch
> (see [build from source instructions](install.md#build-from-source)).

In this benchmarking scenario, we want to measure the effect of in-flight
batch size on token-to-token (T2T) latency. We systematically issue requests to
the server of fixed input sizes and request the model to compute a fixed amount
of tokens in order to increase the in-flight batch size over time.

#### Example

In this benchmark, we will run Perf Analyzer in
[periodic concurrency mode](inference_load_modes.md#periodic-concurrency-mode)
that periodically launches a new concurrent request to the model using

> `--periodic-concurrency-range START END STEP` option.

Reviewer 2: ```suggestion
`--periodic-concurrency-range START:END:STEP` option.
```
